gluster/subvol/README.md

```diff
@@ -0,0 +1,96 @@
+# Gluster Subvol provisioner
+
+## Summary
+
+Use a Gluster Volume with and expose subdirectories as PVCs. Support for
+simplified cloning (based on annotations) is one of the main features.
+
+# Terminology
+
+**subvol**: A PersistentVolumeClaim pointing to a subdirectory on the
+            *supervol*.
+
+**supervol**: A PersistentVolumeClaim used for storing the *subvols*.
+
+**SmartCloning**: Cloning of a PersistentVolumeClaim handled by the provisioner
+		  and storage backend. Requesting a volume to be cloned is done
+                  by passing [Annotations in the PVC](#cloning-by-annotations).
+
+# Features
+
+- dynamically provision a *subvol* through a StorageClass pointing to a *supervol*
+- *supervol* in a different namespace than the *subvol*
+- create/delete Endpoints for each *subvol*
+- mount-options from the *supervol* are copied to the *subvol*
+- delete the subdirectory and contents from the *supervol* upon *subvol* deletion
+- clone other *subvol* PVC from the same *supervol*
+
+# Requirements
+
+- a Gluster Volume that can be used for the *supervol*
+- mounting of *subvol* PVCs require GlusterFS >= 3.13 (released December 2017)
+
+# Installation
+
+- deploy a `gluster-subvol-provisioner` pod
+- create a PVC on a Gluster Volume
+- create a StorageClass that points to the provisioner and the PVC
+
+Examples of the above steps are included in this repository. For deployment
+with a ServiceAccount to allow the needed permissions, see
+[`deploy/gluster-subvol-provisioner-pod.yaml`](deploy/gluster-subvol-provisioner-pod.yaml).
+An example of the StorageClass (with *supervol* PVC) is in
+[`examples/class.yaml`](examples/class.yaml).
+
+# Limitations and Considerations
+
+Currently this provisioner only supports Gluster PVCs. There are very few
+requirements in the code that demand using a Gluster storage environments. With
+minimal changes this provisioner can be made to function with other filesystems
+that provide RWX capabilities for the *supervol*.
+
+## Advantages
+
+- no Gluster management operations
+  - resilent, filesystem operations still function when Gluster servers are down
+  - fast, filesystem operations only
+- can support *SmartCloning* on filesystem level (using reflink, `copy_file_range()`)
+- thin provisioning is automatic (all *subvol* PVCs share the same free space)
+
+## Disadvantages (or TODO)
+
+- no quota support yet, a PVC is not bounded by its requested size
```

Comment:
The intention of the provisioner is to function without Gluster Management operations. The current (to be replaced soonish) Quota support is handled exclusively through the Gluster Management interface. This would introduce a stronger dependency on Gluster, and makes it less likely for this provisioner to become a generic solution that can work with any RWX filesystem.

In addition, quotas add a large performance overhead. It is one of the reasons for it to be replaced in one of the next releases. It may not make sense to implement a feature that needs to be re-written with a next release.

Also, for the current use-case (KubeVirts Containerized-Data-Importer), the contents in the PVC does not grow. The VM disk-image is fully allocated and users of the VM can not exceed the space of the disk-image.

Reply:
openshift.io does not use the Gluster Quota anymore because of scaling issues. Instead they set quotas on XFS on the bricks (hacky, but in their use-case okayish).

Reply:
The size for the PVC request is similar to thin provisioning on Gluster when cloning/snapshots are done. There is no guarantee that the requested size if available.

The main difference is that there is currently no upper-bound enforcement. Quota would address that, but I am not confident that we can recommend users to use it in its current form. Describing the limitation and a TODO for improvement has my preference. Is that acceptable?

Reply:
@nixpanic I still think its (`quota`) good to have as gluster have a mechanism in place. In future, if its deprecated or unsupported we can remove that as well. I still dont know, whats stopping us from coding that effort.  Well, we can survive with documentation but thats the last  option. More or less, the audience of PVC is different than who deploy the provisioner, its unlikely that users themselves apply the limitation in app pod.